File: samples/sample-app-java/src/main/java/com/weedow/searchy/sample/java/repository/PersonDto.java
package com.weedow.searchy.sample.java.repository;

import com.weedow.searchy.sample.java.entity.Person;
import com.weedow.searchy.sample.java.entity.Vehicle;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.rest.core.config.Projection;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Projection(name = "personDto", types = Person.class)
public interface PersonDto {
    String getFirstName();

    String getLastName();

    @Value("#{target.firstName + ' ' + target.lastName}")
    String getFullName();

    String getEmail();

    Set<String> getNickNames();

    Set<String> getPhoneNumbers();

    @Value("#{T(com.weedow.searchy.sample.java.repository.PersonDto).formatAsString(target.getVehicles())}")
    List<String> getVehicles();

    @Value("#{target.addressEntities}")
    Set<AddressDto> getAddresses();

    static List<String> formatAsString(Set<Vehicle> vehicles) {
        return vehicles
                .stream()
                .map(vehicle -> vehicle.getVehicleType() + " - " + vehicle.getBrand() + "/" + vehicle.getModel())
                .collect(Collectors.toList());
    }

}

File: sample-app/src/main/kotlin/com/weedow/spring/data/search/sample/dto/PersonDto.kt
package com.weedow.spring.data.search.sample.dto

data class PersonDto private constructor(
        val firstName: String?,
        val lastName: String?,
        val email: String?,
        val nickNames: Set<String>?,
        val phoneNumbers: Set<String>?,
        val addresses: Set<AddressDto>?,
        val vehicles: Set<VehicleDto>?
) {

    data class Builder(
            private var firstName: String? = null,
            private var lastName: String? = null,
            private var email: String? = null,
            private var nickNames: Set<String>? = null,
            private var phoneNumbers: Set<String>? = null,
            private var addresses: Set<AddressDto>? = null,
            private var vehicles: Set<VehicleDto>? = null
    ) {
        fun firstName(firstName: String) = apply { this.firstName = firstName }
        fun lastName(lastName: String) = apply { this.lastName = lastName }
        fun email(email: String?) = apply { this.email = email }
        fun nickNames(nickNames: Set<String>?) = apply { this.nickNames = nickNames }
        fun phoneNumbers(phoneNumbers: Set<String>?) = apply { this.phoneNumbers = phoneNumbers }
        fun addresses(addresses: Set<AddressDto>?) = apply { this.addresses = addresses }
        fun vehicles(vehicles: Set<VehicleDto>?) = apply { this.vehicles = vehicles }
        fun build() = PersonDto(firstName, lastName, email, nickNames, phoneNumbers, addresses, vehicles)
    }

}